File: rust/75.sort-colors.rs
/*
 * @lc app=leetcode id=75 lang=rust
 *
 * [75] Sort Colors
 */

// @lc code=start
impl Solution {
    pub fn sort_colors(nums: &mut Vec<i32>) {
        let (mut l, mut r) = (0, nums.len() - 1);
        let mut i = l;
        while i <= r {
            match nums[i] {
                0 => {
                    nums.swap(i, l);
                    l += 1;
                    i += 1;
                }
                2 => {
                    nums.swap(i, r);
                    if r < 1 {
                        break;
                    }
                    r -= 1;
                }
                _ => i += 1,
            };
        }
    }
}
// @lc code=end

File: javascript/75.sort-colors.js
/*
 * @lc app=leetcode id=75 lang=javascript
 *
 * [75] Sort Colors
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function (nums) {
  let l = 0,
    i = 0,
    r = nums.length - 1;
  while (i <= r) {
    if (nums[i] === 0) {
      [nums[i], nums[l]] = [nums[l], nums[i]];
      l++;
      i++;
    } else if (nums[i] === 2) {
      [nums[i], nums[r]] = [nums[r], nums[i]];
      r--;
    } else {
      i++;
    }
  }
};
// @lc code=end
